slack_sdk/oauth/installation_store/cacheable_installation_store.py

```diff
@@ -25,12 +27,45 @@ def save(self, installation: Installation):
         return self.underlying.save(installation)
 
     def find_bot(
-        self, *, enterprise_id: Optional[str], team_id: Optional[str]
+        self,
+        *,
+        enterprise_id: Optional[str],
+        team_id: Optional[str],
+        is_enterprise_install: Optional[bool] = False,
     ) -> Optional[Bot]:
+        if is_enterprise_install or team_id is None:
+            team_id = ""
         key = f"{enterprise_id}-{team_id}"
         if key in self.cached_bots:
             return self.cached_bots[key]
-        bot = self.underlying.find_bot(enterprise_id=enterprise_id, team_id=team_id)
+        bot = self.underlying.find_bot(
+            enterprise_id=enterprise_id,
+            team_id=team_id,
+            is_enterprise_install=is_enterprise_install,
+        )
         if bot:
             self.cached_bots[key] = bot
         return bot
+
+    def find_installation(
+        self,
+        *,
+        enterprise_id: Optional[str],
+        team_id: Optional[str],
+        user_id: Optional[str] = None,
+        is_enterprise_install: Optional[bool] = False,
+    ) -> Optional[Installation]:
+        if is_enterprise_install or team_id is None:
+            team_id = ""
+        key = f"{enterprise_id}-{team_id}={user_id}"
+        if key in self.cached_installations:
+            return self.cached_installations[key]
+        installation = self.underlying.find_installation(
```

Comment:
what does `self.underlying` do here?

Reply:
The `underlying` can be `FileInstallationStore` or whatever implements the `InstallationStore` methods. See the constructor.